pkg/services/datasources/models.go

```diff
@@ -149,6 +149,12 @@ type DeleteDataSourceCommand struct {
 	DeletedDatasourcesCount int64
 
 	UpdateSecretFn UpdateSecretFn
+
+	// Optional way to skip publishing delete event for data sources that are
+	// deleted just to be re-created with the same UID during provisioning.
+	// In such case we don't want to publish the event that triggers clean-up
+	// of related resources (like correlations)
+	SkipPublish bool
```

Comment:
Not sure this will work. What if there are other consumers besides correlations relying on these events being published? Correlations delete handler somehow needs to make the decision to handle or not - can it check if the datasource is read only and skip it as an example?

Is there a problem correlations being deleted and re-created? Would simplify the code to separate/break datasource and correlation provisioning in two distinct packages/provisioners

Reply:
> What if there are other consumers besides correlations relying on these events being published? 

The idea is that this is set only during provisioning and only when deleting data sources that will be re-created. New resources don't need to worry about this - the `DataSourceDeleted` won't be published at all so the handler won't be triggered. This happens only for data sources that are not truly deleted, but re-created. For example it is used in Cloud to rename a stack.

> Correlations delete handler somehow needs to make the decision to handle or not (...)

That would mean that every new consumer would need to handle it. There's a problem that we only know if a data source was read-only, not if it was provisioned. We could pass this information with `DeleteDataSource` event instead of guarding this with `SkipPublish`. However in our case, it's not a problem to have different logic during provisioning or not - it's that we want to perform clean-up only if the data source is actually deleted, not-recreated.

> Is there a problem correlations being deleted and re-created?

The main problem is that when a data source is removed we want to remove all created correlations (provisioned or done manually). However in Cloud, when the stack is renamed, a data source is removed and created once again through provisioning. Without `SkipPublish` all user-created correlations would be removed.

> Would simplify the code to separate/break datasource and correlation provisioning in two distinct packages/provisioners

You mean to have separate handlers for deletion happening during provisioning or not? In theory, as mentioned above, instead of `SkipPublish` we could always publish the event, but basically pass a boolean indicating if that happened during provisioning or not and let the handler decide if only provisioned resources should be removed or not. But then we end up in a similar problem as described above, but the other way round. If a data source is actually removed during provisioning, we also want to remove user-created correlations. 


I agree it's far from ideal. The main use case we want to cover is renaming the stack in Cloud. If that wasn't handled by deleting and creating datasources we could argue that if a data source is deleted in provisioning, it should have all resources removed even if it's re-created. But we know it's not fully true, and will create problems for Cloud customers.

Reply:
@marefr - does it make sense? Do you think we should take a different approach here?

Reply:
Sorry for slow response. I'll try to come back with a response shortly

Reply:
Got it I see the challenge now. Then I probably don't see any other alternative than your suggested approach or stop caring that datasources are deleted as dashboards does it.

> Another reason is that DeleteDataSource event was added for Correlation in https://github.com/grafana/grafana/pull/51630, so exiting resources do not use it, and AFAIK there's no other resources using it.

If someone starts using it then it might be an unexpected behavior that it's not triggered on provisioned datasources and it's very much hidden that this behavior exists.

Reply:
> If someone starts using it then it might be an unexpected behavior that it's not triggered on provisioned datasources and it's very much hidden that this behavior exists.

The skipping is applied only in one case: when someone deletes and creates a data source with the same UID during single provisioning process. If the user deletes the data source by adding it to `deleteDatasources` list without creating a data source with the same uid - it will be deleted as normal.

I agree it's not clear this behavior exists. However, without it we end up with another unexpected behavior: "You can change your data source name by deleting and re-adding it, but... you will lose all resources related to your data source". 

I guess a better solution could be if we had an option in provisioning to rename a data source. Maybe that's the way to follow-up on this? With something like:

```
renameDatasources:
  - previous: Previous Name
     new: New Name
     uid: UID
```

We could use it in the Cloud to rename stacks and get rid of that event skipping.

Reply:
> I guess a better solution could be if we had an option in provisioning to rename a data source. Maybe that's the way to follow-up on this? With something like:

Sorry probably confused again. Unless you rename a datasource UID (which shouldn't happen) any correlations created referencing this datasource uid should continue to function given that you don't delete correlations when a datasource is deleted. So maybe what I'm suggesting is to not delete correlations when datasource is deleted. Doesn't matter I guess since we've already discussed this and you landed in this solution.